dao/FashionKnex.js

```diff
@@ -27,35 +27,66 @@ class FashionKnex {
     this._knex = knexConnection || knex(knexConfig)
   }
 
-  async insert (data, returning) {
+  /**
+   * TODO: replace with https://github.com/tgriesser/knex/pull/2197 after new
+   * and improved implementation is submitted
+   * Ref: https://gist.github.com/plurch/118721c2216f77640232
+   * http://www.postgresql.org/docs/9.5/static/sql-insert.html
+   */
+  async upsert (data, {conflictColumn = 'id', returning = '*'} = {}) {
     data = _clean(data)
-    returning = returning || '*'
 
-    let result
+    const insertString = this._knex
+      .insert(data)
+      .into(this._tableName)
+      .toString()
+    const conflictString = this._knex.raw(` ON CONFLICT (??) DO NOTHING`,
+      conflictColumn).toString()
+    const query = (insertString + conflictString + ` RETURNING ${returning}`)
+      .replace(/\?/g, '\\?')
+
+    return this._knex.raw(query)
+      .then(result => result.rows[0]) // rows is an empty array if pre-existing
+      .catch(err => {
+        this._logger.error('Error during upsert', err)
+        throw err
+      })
+  }
 
-    try {
-      result = await this._knex
-        .insert(data, 'id')
-        .into(this._tableName)
-        .returning(returning)
-    } catch (err) {
-      const stringified = JSON.stringify(data)
-      this._logger.error(`Error inserting data "${stringified}"`, err)
-      throw err
-    }
-    return result
+  async batchInsert (data, {returning = '*'} = {}) {
+    const dataArray = data.map(row => _clean(row))
+
+    return this._knex
+      .batchInsert(this._tableName, dataArray)
+      .returning(returning)
+      .catch(err => {
+        this._logger.error('Error during batch insert', err)
+        throw err
+      })
+  }
+
+  async insert (data, {returning = '*'} = {}) {
+    data = _clean(data)
+
+    return this._knex
```

Comment:
`async/await`?

Reply:
ditto

Reply:
I'm already using `async` here - I just return the promise here. I could `return await` but that seems unnecessary since synchronous errors get captured in `async` functions and result in a rejected promise anyway. I think the real distinction is whether to use a try/catch versus .catch chaining. I only used `.catch` here because I don't need to modify the result returned by the promise or do anything complex on the error path. I don't think this is a particularly uncommon pattern, but I might could be convinced if you think otherwise.